File: leetcode/stack/min_stack.py
class MinStack:
    def __init__(self):
        self.stack = []
        self.min_stack = []

    def push(self, val: int) -> None:
        self.stack.append(val)
        current_min = val if not self.min_stack else min(val, self.min_stack[-1])
        self.min_stack.append(current_min)

    def pop(self) -> None:
        self.stack.pop()
        self.min_stack.pop()

    def top(self) -> int:
        return self.stack[-1]

    def getMin(self) -> int:
        return self.min_stack[-1]

# Try it out — press Run
ms = MinStack()
ms.push(1)
ms.push(2)
ms.push(0)
print(ms.getMin())  # 0
ms.pop()
print(ms.top())     # 2
print(ms.getMin())  # 1

File: leetcode/stack/min_stack.ts
class MinStack {
  private stack: number[] = [];
  private minStack: number[] = [];

  push(val: number): void {
    this.stack.push(val);
    const currentMin = this.minStack.length === 0
      ? val
      : Math.min(val, this.minStack[this.minStack.length - 1]);
    this.minStack.push(currentMin);
  }

  pop(): void {
    this.stack.pop();
    this.minStack.pop();
  }

  top(): number {
    return this.stack[this.stack.length - 1];
  }

  getMin(): number {
    return this.minStack[this.minStack.length - 1];
  }
}

// Try it out — press Run
const ms = new MinStack();
ms.push(1);
ms.push(2);
ms.push(0);
console.log(ms.getMin()); // 0
ms.pop();
console.log(ms.top());    // 2
console.log(ms.getMin()); // 1

export {};
